lib/cache.js

```diff
@@ -4,7 +4,61 @@ const log = require('npmlog')
 const pacote = require('pacote')
 const path = require('path')
 const rimraf = promisify(require('rimraf'))
+const semver = require('semver')
 const BaseCommand = require('./base-command.js')
+const npa = require('npm-package-arg')
+const jsonParse = require('json-parse-even-better-errors')
+
+const searchCachePackage = async (path, spec, cacheKeys) => {
+  const parsed = npa(spec)
+  if (parsed.rawSpec !== '' && parsed.type === 'tag')
+    throw new Error(`Cannot list cache keys for a tagged package.`)
+  const searchMFH = new RegExp(`^make-fetch-happen:request-cache:.*(?<!/[@a-zA-Z]+)/${parsed.name}/-/(${parsed.name}[^/]+.tgz)$`)
+  const searchPack = new RegExp(`^make-fetch-happen:request-cache:.*/${parsed.escapedName}$`)
+  const results = new Set()
+  cacheKeys = new Set(cacheKeys)
+  for (const key of cacheKeys) {
+    // match on the public key registry url format
+    if (searchMFH.test(key)) {
+      // extract the version from the filename
+      const filename = key.match(searchMFH)[1]
+      const noExt = filename.slice(0, -4)
+      const noScope = `${parsed.name.split('/').pop()}-`
+      const ver = noExt.slice(noScope.length)
+      if (semver.satisfies(ver, parsed.rawSpec))
+        results.add(key)
+      continue
+    }
+    // is this key a packument?
+    if (!searchPack.test(key))
+      continue
+
+    results.add(key)
+    let packument, details
+    try {
+      details = await cacache.get(path, key)
+      packument = jsonParse(details.data)
+    } catch (_) {
+      // if we couldn't parse the packument, abort
+      /* istanbul ignore next */
+      continue
+    }
+    /* istanbul ignore next */
+    if (!packument.versions || typeof packument.versions !== 'object')
+      continue
+    // assuming this is a packument
+    for (const ver of Object.keys(packument.versions)) {
+      if (semver.satisfies(ver, parsed.rawSpec)) {
+        /* istanbul ignore else */
+        if (packument.versions[ver].dist
+          && typeof packument.versions[ver].dist === 'object'
```

Comment:
so we've verified that `packument.versions[ver].dist` is an object, but what if that object doesn't have a `tarball` property? is that worth checking for here? the consequence is we'll search for `make-fetch-happen:request-cache:undefined` as a key which _could_ hit a match, but if it does that would be surprising and most likely indicative of a bug somewhere, so maybe it's ok to ignore it?

Reply:
addressed by adding a condition